### ISSUE TITLE
Transition to Java 8 (was 32- vs 64-bit java)

### ISSUE THREAD

**Issue by kdahlquist** [lmu-bioinformatics/xmlpipedb#3]
So, previous versions of GenMAPP Builder could only run an export using 32-bit java, not 64-bit java.  You mentioned that java 8 is now 64-bit, so I downloaded and installed the 64-bit jdk.  The filename for the installer is:  jdk-8u51-windows-x64.exe.  I just finished the import and processing of the UniProt, GO, and GOA source files and tried to start an export.  I got a warning message saying that it can't be done in a 64-bit environment and to download and install the 32-bit java.  I will do this now, but I wanted to clarify with @dondi that it is still the case that exports must be done in 32-bit.

**Comment by kdahlquist** [comment]
An update:  I installed 32-bit java "jdk-8u51-windows-i586.exe", corrected the gmbuilder-32bit.bat file to read:

@echo off

"C:\Program Files (x86)\Java\jre1.8.0_51\bin\java" -D32 -Xmx1024m -jar gmbuilder.jar

launched gmb2b84 and attempted to start an export.  It seems to start, but I noticed an error in the console window that indicated that it did not find a database connection.  I checked the database configuration window again and it seemed to be right and tried again and got the same error.

The screenshot is on the SourceForge wiki page here:

http://xmlpipedb.sourceforge.net/wiki/index.php/Gene_Database_Testing_Report_V._cholerae_2015-08-10#Export_Information

If you can get back to me by Tuesday afternoon, I can try again before Wednesday's meeting.

**Comment by dondi** [comment]
I looked into this issue and the core piece here is the so-called "database driver"—the part of the software that allows Java to communicate with a database system. Import is not an issue because the PostgreSQL driver is pure Java and thus we can include it with our distribution. Export has been what has kept GenMAPP Builder as a Windows-only application, because the *Access* database driver had been Windows-only. (it had a Java layer, which was how Java communicated with it, but its main functionality is Windows-native)

So the specific issue to address here is getting the Access database driver on that machine. We never thought about it before because it used to be available by default. It seems that it is optional on newer versions of Windows.

This StackOverflow Q&A appears to have the most potential:

http://stackoverflow.com/questions/19334723/how-to-connect-to-access-mdb-database-from-64-bit-java

Extracting from this page, I would try the following first:

* Follow the download link to the 64-bit version of the driver (here it is copied out: http://www.microsoft.com/en-US/download/details.aspx?id=13255)

This does require a change to the code, so I will make a provisional build and supply it in a draft release.

The UCanAccess option is intriguing, and potentially decouples GenMAPP Builder from Windows (i.e., we may be able to perform exports from Linux and/or Macs with this one). But it requires testing: http://ucanaccess.sourceforge.net/site.html

**Comment by dondi** [comment]
OK, I've prepared a draft release 2.0b85 here: https://github.com/lmu-bioinformatics/xmlpipedb/releases/tag/untagged-24f74e935b0ccdb6eb2a

Use the instructions in *this* comment to supersede the ones above. Proceed in the order given, continuing to the next alternative only if the current one does not get you past this driver issue:

1. First, try this draft release on its own. If the 64-bit driver is already on your computer, the modifications to the code (taken from http://www.microsoft.com/en-US/download/details.aspx?id=13255) should now "connect" GenMAPP Builder to that driver, and export should be able to proceed beyond that point.

2. If an error that looks similar to the one you encountered still comes up (i.e., it is looking for the `OdbcJdbcDriver` or something similar), then perhaps the driver is not installed. In that case, go ahead and perform a download + install of the package given in http://www.microsoft.com/en-US/download/details.aspx?id=13255 (same link as above).

3. Try running this release with 32-bit Java 8.

If none of these alternatives work...then there is more to do. But we can cross that bridge if we get there.

**Comment by kdahlquist** [comment]
Option 1 of using the draft release with the 64-bit java did not work.  Screenshot of error available here.

http://xmlpipedb.sourceforge.net/wiki/index.php/Gene_Database_Testing_Report_V._cholerae_2015-08-10#2015-08-11

Moving on to option 2.

**Comment by kdahlquist** [comment]
The link for the driver has options to download either a 32-bit or a 64-bit driver.  I guess I'll try the 64-bit first.

**Comment by dondi** [comment]
I read up on this further and encountered new information—you can stop at this point, because I have discovered that ODBC bridge support was removed in Java 8. Our remaining options are thus: switch to UCanAccess, or revert to Java 7. Sorry I did not spot this one earlier (links forthcoming).

**Comment by dondi** [comment]
Oops sorry for the misclick. Here is the information from Oracle:

https://docs.oracle.com/javase/7/docs/technotes/guides/jdbc/bridge.html

**Comment by dondi** [comment]
...and instructions for UCanAccess, if we go that route. I can prepare a build that follows these instructions and we can see how well that works:

http://stackoverflow.com/questions/21955256/manipulating-an-access-database-from-java-without-odbc

**Comment by kdahlquist** [comment]
It seems that UCanAccess is the better route since it doesn't involve the installation of a driver by the user, is that the case?  I like this better than reverting to Java 7 since we can expect that we will get very out of date with java very quickly this way.

**Comment by dondi** [comment]
Yes, UCanAccess can be included in our distribution and no other installation will be necessary outside of Java. OK, I will work on putting this together and release something. If all goes well I may have something by our meeting.

**Comment by dondi** [comment]
OK, good news and bad news:
- Good news—I have successfully integrated UCanAccess into the codebase.
- Good news—As promised, UCanAccess is pure Java and thus I am able to initiate an export on an operating system other than Windows (e.g., Mac OS X).
- Bad news—There are some differences between the SQL that is accepted by the original database driver vs. UCanAccess. Porting things over has been a matter of starting an export, looking for when an error comes up, then adjusting the query being asked at that time. These adjustments (so far) have ranged from brain-dead easy to requiring some refactoring.

As a result of this, I was not able to create a release update and will have to continue working on this tomorrow. After this is done, there might be some lingering issues. One that I can already foresee is that UCanAccess does not appear to accept hyphens in table names (e.g., `UniProt-GoCount`). This is not completely confirmed yet, but as of now this is a possibility.

More details in the meeting, but this is the latest so far.

**Comment by kdahlquist** [comment]
At 01:25 AM 8/19/2015, @dondi  wrote:

Yeah, it is safe to say this has not been plug and play :-\ Issues remain but I decided to do an intermediate draft release so that (if you have the chance) there can at least be independent confirmation of the progress that I did make so far. The draft release on GitHub has the .zip, along with a description of the remaining issues. Despite those, I'm still hoping to verify that the other adjustments that I made and am seeing can also be seen on your end. This will at least assure me that those changes can stick.

Details at our meeting if you care to hear them...I'm hoping that these remaining issues can be solved soon.

**Comment by kdahlquist** [comment]
At 02:28 AM 8/20/2015, @dondi  wrote:

I managed to get the database inserts working right (as far as I can tell). No errors reported, no empty tables. Phew! This leaves the underscores in the names vs. hyphens. Last hurdle...

Meanwhile, I have updated the .zip file in the draft release, if you want to give that one a shot. No worries either if you prefer to wait for the restoration of hyphens in the relationship name

**Comment by kdahlquist** [comment]
Quick reminder to @dondi to use the GitHub issues for correspondence regarding GenMAPP Builder.  I pasted your last two e-mails to me as comments above.

The Vibrio export I began yesterday (8/19/15) at 11:25:46 AM ended at 7:42:27 PM about 8 hours later.  The time length might have been because my computer was busy doing other things at the same time, I can't comment on that.

I believe you told me yesterday that you had seen empty tables with your export of Shewanella.  My export of Vibrio finished with all tables populated and the OriginalRowCounts accurately giving the table counts.  I'm afraid that you might have been chasing a Shewanella-specific bug.  I will go ahead and start another export with the current draft release, it's no problem to have that run in the background.

**Comment by dondi** [comment]
Thanks for the updates—and also for the reminder to just go here :) FWIW the latest draft release does populate all tables, including Shewanella, so at least it is no longer an issue either way. I left things this morning with a new build doing an export that seeks to restore hyphens in relationship tables rather than underscores. If that resolves, then I think we can finally make the release public.

**Comment by kdahlquist** [comment]
Any updates?

**Comment by dondi** [comment]
Yes, but unfortunately none that led to a release yet. It turned out that restoring hyphens to the GO-related export code was a bigger task than expected. Going back to hyphens has meant bypassing the limited SQL interpretation capabilities of UCanAccess and going straight to the direct table manipulation of Jackcess. GO export performs both queries and inserts, and the queries are what has taken a while. Essentially any `select ... from ... where ...` clause that involves a table with a hyphen has needed to be translated into their direct code equivalent. I knew those were lurking but there have been more than anticipated. I know of a couple more and will get to those tonight. I was hoping not to spill over the weekend but that's how the chips fell unfortunately.

**Comment by dondi** [comment]
OK, for the first time, I can say that we have a draft release with “no known issues.” Hyphens are back and restored, tables are populated, and GO is now on the new direct Jackcess code as well.

I’ve now renamed/renumbered the release, alongside its writeup:

https://github.com/lmu-bioinformatics/xmlpipedb/releases/edit/untagged-e30da84e8544a0dfb024

I tested exports in both *Shewanella* and *Vibrio* (the latter used fresh downloads from today). No errors observed, and eyeball inspection of all tables showed expected values. In the case of *V. cholerae*, the double-entry customization (with underscore and without underscore) was seen in the resulting export.

In other scenarios I would actually release this—however given the depth of changes involved, I’m more inclined for one last round of independent confirmation before doing the final merge-and-tag. Hope it works well on your end!

**Comment by kdahlquist** [comment]
Can you post the exact source files you used for Vibrio so I can compare my results with yours?  It would have been helpful if you had used the Vibrio files I put on Lionshare so I didn't have to reimport everything to compare my export with yours.  Thanks,


Kam D. Dahlquist, Ph.D.
Associate Professor of Biology
Loyola Marymount University
http://myweb.lmu.edu/kdahlqui



-------- Original message --------

**Comment by dondi** [comment]
Ah—that’s what I recalled from our meeting, that you had sent me those files, but when I checked the email from LionShare I only saw the .gdb in the link; I didn't login to LionShare directly. I’m uploading the files I used and will link to them when done.

**Comment by dondi** [comment]
Hmmm, now that I am logged in I don’t see how to access the source files. The original LionShare email links to the gdb only; I tried it anyway in case it was actually a folder or .zip file, but I did just get the .gdb. In any case, files are uploaded and shared; to be safe I shared them individually.

**Comment by kdahlquist** [comment]
I just finished an export with the existing Postgres database I made with the versions of the source files I used before (from 8-10-15, http://xmlpipedb.sourceforge.net/wiki/index.php/Gene_Database_Testing_Report_V._cholerae_2015-08-10).

It finished kind of fast, and the reason for that was that the Gene Ontology table had zero records and the GeneOntologyCount had 3 and GeneOntologyTree had 4.  Interestingly, the other *-GeneOntology tables were populated.

 A visual inspection of the rest of the tables seems OK.

 I can go ahead and run an export with the source files you Lionshared to me today to see if there is any difference.

**Comment by kdahlquist** [comment]
I just started the export based on the source files you sent via lionshare.  I can already tell that something is different because with the previous export the progress bar jumped to 65% almost immediately, but now it's on 1% where it should be.

So, it looks like something is different about the GO data.  I noticed that you are using a GO file from June and I am using one from August...

**Comment by dondi** [comment]
Yes, I didn’t redownload the GO data from June when I started troubleshooting *Shawnella* but apparently a breaking change has appeared in the August version. I will report what I see when I do a cycle with the source file links from today.

**Comment by dondi** [comment]
I’ve done an export with the source files from LionShare and I’ve subsequently shared the resulting .gdb with you (hope the email showed up OK). I started with an empty database, then loaded UniProt and GO OBO. Upon loading GO OBO I went ahead and did the processing. Finally I loaded the GOA, then started the export.

The resulting .gdb did have GeneOntology entries—6252 in fact. So that is a definite difference in result. Eyeball inspection yielded no obvious anomalies.

To be certain, I did a second export from nothing, using the source files again, and got files of the same length. I spot-checked a few tables and they looked the same also.

Thus, things look good except for the empty GeneOntology tables. As one last check, I looked at the 20150819 gdb that was shared via LionShare, and that .gdb had 6252 records, like the exports performed today. (although at that time, table names still had underscores)

Take a gander and let me know what you find...

**Comment by kdahlquist** [comment]
So, to be clear, you used the 8/10/15 source files from LionShare and got a complete gdb?  I'll start over from scratch again to see what I get.

One weird thing, I left an export running overnight when I left yesterday with the other set of Vibrio files you sent me with the June GO in them.  When I got in this morning, it had not finished and was still stuck at 1%, so I canceled the export.

Like I said, I'll try again from scratch and see what I get.

**Comment by dondi** [comment]
Affirmative: I started from scratch and used the sources from LionShare. PostgreSQL 9.4.x (I'm not recalling the x right now), Java 1.8u60, and the draft GenMAPP Builder 3.0 release. For verification I did it twice. Resulting .gdb's showed the same information on spot-check, and had the exact file size.

That is interesting about the June GO export...I wonder if that is a clue on what might be different. Let me know what you see after starting from scratch.

**Comment by kdahlquist** [comment]
So, I think the export has stalled at the same point as the one I tried last night.  I forgot to take a screenshot of the console of the previous stall, but I have posted one of the current stall.  I have cancelled the export yet, just in case.  http://xmlpipedb.sourceforge.net/wiki/index.php/Gene_Database_Testing_Report_V._cholerae_2015-08-25#Export_Information

I am now trying another fresh import/export cycle with the 8/10/15 source data on my laptop (the previous was on one of my lab computers) just to see if something is different.

**Comment by dondi** [comment]
Hmmmm...the stall with no errors is concerning. One thing to try—after cancelling this, the nascent .gdb is still there, yet? From the log it looks like it stalled on GeneOntologyTree. Try opening the incomplete .gdb and see what is in GeneOntologyTree, whether it is created/empty/partially with records. This may at least tell us how far it got.

We’ll see what happens with the laptop next...

**Comment by kdahlquist** [comment]
I opened the one that stalled last night and cancelled the one that stalled now on the lab machine and they both have the following:

GeneOntology: populated
GeneOntologyTree: populated
UniProt-GeneOntology: populated
GOCount: empty
UniProt-GOCount: empty

The laptop export has now reached this point as well, but I haven't cancelled it yet.

I have to admit I haven't been keeping as good a track as I could of the exports.  But I'm pretty sure that the last export that completed for me was using a Postgres database that I had built with a previous version of gmbuilder.

**Comment by kdahlquist** [comment]
I have a further error in the console window to show.  It is posted on http://xmlpipedb.sourceforge.net/wiki/index.php/Gene_Database_Testing_Report_V._cholerae_2015-08-25#Export_Information

However, it may have been caused by me opening the gdb in Access without cancelling the export first.

**Comment by dondi** [comment]
OK, all of that information makes sense; GOCount and UniProt-GOCount are the last of those tables that are populated in the code and if things got stuck at GeneOntologyTree then it makes sense that those tables are empty. How many records are in GeneOntologyTree? My exports have 122893.

The lack of feedback, even an error message, makes this tricky (let alone the fact that I am unable to replicate this behavior). Thus, I prepared a new build whose only difference is additional instrumentation for this part of the code. I've labeled it 3.0b2 and have uploaded it (we'll consider the last version actually 3.0b1 even if I did not label it as such—it was a draft release anyway).

The main new log is that the GeneOntologyTree output will emit a message every 250 records. This will give us a signal of where things stall—or whether things are just really slow. If the former, then maybe something will be indicated based on when it stalls (i.e., immediately vs. near the end). If the latter…well at least it doesn’t stall, but if it is unacceptably slow then that is a problem too.

Oh also, my exact PostgreSQL version is indeed 9.4.0, and I am at Java 1.8.0_60. Let's see what the extra logging reveals...

**Comment by kdahlquist** [comment]
The overnight run on my laptop never finished, I will proceed to try the new draft release.

**Comment by kdahlquist** [comment]
@dondi, question: is your version of Postgres 9.4.0 or 9.4.4?  On the download page there is no installer for 9.4.0, it goes from 9.3.9 to 9.4.4. http://www.enterprisedb.com/products-services-training/pgdownload#windows

Is update 60 of java really new?  I thought I had downloaded the latest installers when I was installing software on my laptop and I only had update 51.

I'm going to try to get one computer to use exactly the versions you are using and leave the other one alone for this next test.

**Comment by kdahlquist** [comment]
More information, from the previous export that stalled, the GeneOntologyTree table had 17355 records.  The last record is:

OrderNo	Level	ID	Name
17355	11	0044206	UMP salvage

I have exports running on two machines with the extra logging, I'll let you know when there is something to report.

**Comment by dondi** [comment]
OK, I will look at my exports to see what record 17355 or thereabouts looks like. And yes let me know how the logged exports look. Based on the rate of 250 records per log statement, total time can be estimated based on the 122000 or so that I got.

**Comment by kdahlquist** [comment]
I'm leaving at the end of the day here and both exports have passed the 17355 mark.

I will coming in at noon for our meeting tomorrow, so you'll know as soon as I do what happens.

**Comment by dondi** [comment]
Hmmmmm.....I wonder then if what we are seeing then is a serious slowdown in this particular export step due to my refactoring. If we do not see a stall by the meeting (i.e., records still seem to be exported 250 at a time, but just very slowly), then I can revisit the code with a closer eye on speed.

**Comment by dondi** [comment]
BTW my exported .gdb has the same information at record 17355. So at least they match up at that point.
<img width="1364" alt="vc-gene-ontology-tree-20150824" src="https://cloud.githubusercontent.com/assets/382242/9515842/21c49a02-4c59-11e5-8fa3-2323b17dce9d.png">

**Comment by dondi** [comment]
As a “preemptive release" in case the issue does turn out to be a severe performance degrade due to the change in database implementation, I have uploaded 3.0b3 to use the previous code. The original change in implementation was motivated by the need to have tables with hyphens in their names; fortunately, a revert is possible here, because GeneOntologyTree does not have hyphens. It happened to "come along for the ride" to keep things consistent, but in exchange for better performance we may need to support both approaches.

No need to jump on this just yet; we can wait on what the overnight exports reveal at our meeting.

**Comment by kdahlquist** [comment]
Laptop is at 88250 and the lab computer is at 99500.  I just saw the laptop increment, so it looks like it is very slow as opposed to stalled.  We'll discuss what to do in a few minutes when you get here.

**Comment by dondi** [comment]
As discussed in today’s meeting, I have reverted GO table code for tables without hyphens in their names to the original UCanAccess implementation—this is now 3.0.0-build-4. There should be no functional difference other than speed. The system tables, it turned out, have a structure that makes it harder to pick one database implementation approach over the other. That may have to be a later speed tweak.

I was not able to get to the *Date* column issue in the Systems table; I can look into that for a future build.

**Comment by kdahlquist** [comment]
So, both computers finished around the same time last night, but with drastically different elapsed times.

laptop with build 3 finished at 9:18:57 PM with an elapsed time of approximately 9 hours

boulardii with build 2 finished at 9:54:01 PM with an elapsed time of approximately 32 hours

I will update once I've had a chance to take a look at the actual gdb's.  

I will start a new export with build 4.

I'll post the date issue as a separate issue so as not to muddy the waters.

**Comment by kdahlquist** [comment]
An export with build 4 finished in a blazing hour and 15 minutes on boulardii.  The OriginalRowCounts table is identical to the exports from yesterday.

I'm in the process of vetting the gdb, hence the other issues (#9, #10, #11).  Things are looking good so far.  A full report will be forthcoming on the SourceForge wiki.

**Comment by dondi** [comment]
Given that we have successfully exported from this branch in class, I think it is safe to close this issue, alongside the related #9, #10, and #11. That doesn’t mean there aren’t more bugs lurking, but in terms of the transition I think it is complete.

Accordingly, with closing this issue I think we can also merge this branch into `master`. Again, not saying it’s perfect, but if new stuff arises, those would be new issues in their own right.

Merging into `master` will also simplify the startup processes for the gene database projects—it was in writing up those pages that I realized it is probably the right time to do this.

Anyway, I figured I’d stop and sanity-check this with an issue comment in case I’m missing some detail that calls for *not* merging yet. Let me know what you think; thanks!
